### PR TITLE
Refacto note service

### LINKED ISSUE #5: Améliorer l’architecture du service de note
### Description

En tant que développeur j'aimerai pouvoir vite comprendre l'architecture du projet et développer des features plus rapidement

### Definition of Done

- [ ] Dans le service de note, modifier la fonction  Create en Base de donnée pour stocker les blocks indépendamment des notes
- [ ] Dans le service de note, modifier la fonction  Update en Base de donnée pour accéder et modifier les blocks indépendamment des notes
- [ ] Dans le service de note, modifier la fonction  Get en Base de donnée pour accéder aux blocs en fonction de leur id
- [ ] Dans le service de note, modifier le message AddBlock et rajouter le champs noteId pour savoir a quelle note appartient quel bloc
- [ ] Dans le service de note, modifier le message UpdateBlock et rajouter le champs noteId pour savoir a quelle note appartient quel bloc
- [ ] Dans le service de note, appeler le middleware du package Auth dans le service
- [ ] Implémenter les routes dans l’API Gateway qui appellent les methodes du service de note

### Estimation

Estimée à **6** JH.

### LINKED ISSUE #6: Adapter le service a protobuf style v2
### Description

En tant que développeur je veux pouvoir echanger des messages a l'api gateway

### Definition of Done

- [x] Modifier le .gitsubmodules
- [x] Regénérer le protorepo dans mon service en tant que submodule
- [x] Changer les imports des fichiers .go du service

### Estimation

Estimée à **2** JH.

### LINKED ISSUE #7: Finir les tâches du dernier sprint
### Description

Implémenter avec notre nouvelle architecture de projet le CRUD des blocs

### Definition of Done

- [x]  Créer les modèles `Block`, `BlockWithIndex` et `FilterBlocs` dans le dossier models/ de mon projet
- [x]  Créer les fonction `Update`, `Create` et `Delete` pour ajouter, modifier, et supprimer un bloc en base de donnée

### Estimation

Estimée à **2.5** JH.

### PR DESCRIPTION
**Pull request by gabriel654165**
#### Description

Environment variables added in main.go for :
- environment (prod or dev)
- port 
- mongoDbUri (noted database)
- mongoDbName ("notes-service")
- jwtPrivateKey (in order to use the auth package)
- NoteDatabase and NoteCollection (are going to be use in mongo/ methods and initialized in server.go)

Also protobuf style v2 implemented in notes.go, mongo/notes.go and blocks.go, mongo/blocks.go

All methods are returuning ```status.Errorf(codes.Unimplemented, "not implemented")```
In the next PR all methods and enpoints are going to be implemented. But for readable purpose I make this PR first and all the other are going to be based on this PR.

Unit test done and working for empty methods

Fixes #5 
Fixes #6 
Fixes #7 

#### Changelog

- Protobuf style v2 implemented
- Dependancy injection of logger, notes or blocks repository
- Grpc api server

### CODE REVIEW

File: models/mongo/blocks.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+)
+
+type block struct {
+	ID      string `json:"id" bson:"_id,omitempty"`
+	NoteId  string `json:"noteId" bson:"noteId,omitempty"`
+	Type    uint32 `json:"type" bson:"type,omitempty"`
+	Content string `json:"content" bson:"content,omitempty"`
+}
+
+type blockWithIndex struct {
```
Comment (maxime-dodin) [inline models/mongo/blocks.go:20]: dans quelle cas on vas utiliser le BlockWithIndex à la place du Block ?

**Review by maxime-dodin — APPROVED** [review]

File: .envServer [hunk 2]
```diff
@@ -0,0 +1,3 @@
+PORT=3000
+MONGO_PASSWORD=Gyy628\\nAWS
+MONGOO_URI=mongodb://localhost:27017
```
Comment (rojas-diego) [inline .envServer:3]: I'm not sure we should commit the `.env`. Besides, even if these are just for development, let's not commit and push database credentials. WDYT?

File: Dockerfile [hunk 3]
```diff
@@ -11,7 +11,6 @@ RUN apk update && apk add --no-cache make protobuf-dev=3.18.1-r1
 RUN go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
 RUN go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
 RUN ./misc/gen_proto.sh
-
 RUN go build -buildvcs=false .
 
 FROM alpine:latest
```
Comment (rojas-diego) [inline Dockerfile:16]: This Dockerfile is outdated/invalid. We don't generate protobuf manually anymore. Refer to the Dockerfile from the accounts service.

File: Makefile [hunk 4]
```diff
@@ -1,34 +1,49 @@
...
 build:
 	go build .
-# Create an executable for service note
-re:
+
+init:
 	rm go.mod
-	go mod init notes-service
+	go mod init ${EXE}
 	go mod tidy
```
Comment (rojas-diego) [inline Makefile:17]: I don't see the goal of this rule. `go mod init` should be run once and never again?

File: Makefile [hunk 5]
```diff
@@ -1,34 +1,49 @@
...
+########### !Project rules #########
+
+########### Submodules rules ##########
+
 codegen: update-submodules
-	docker run --rm -v `pwd`/grpc:/app/grpc -v `pwd`/misc:/app/misc -w /app noted-go-protoc /bin/sh -c misc/gen_proto.sh
-# Run the golangci-lint linter.
-lint:
-	docker run -w /app -v `pwd`:/app:ro golangci/golangci-lint golangci-lint run
+	docker run --rm -v `pwd`/protorepo:/app/protorepo -v `pwd`/misc:/app/misc -w /app noted-go-protoc /bin/sh -c misc/gen_proto.sh
```
Comment (rojas-diego) [inline Makefile:29]: This rule is outdated. We don't generate protobufs manually anymore.

File: Makefile [hunk 6]
```diff
@@ -1,34 +1,49 @@
...
-# Fetch the latest version of the protos submodule.
 update-submodules:
-	git submodule update --recursive grpc/protos
 	git submodule update --remote
-#	git pull --recurse-submodules
-# After cloning the repo, run init
+
 init-submodules:
 	git submodule init .gitmodules
 	docker build -t noted-go-protoc -f misc/Dockerfile .
```
Comment (rojas-diego) [inline Makefile:36]: Same here.

File: Makefile [hunk 7]
```diff
@@ -1,34 +1,49 @@
...
 
-## DataBase rules
-# no rules we have a oncline cluster
+lint:
+	docker run -w /app -v `pwd`:/app:ro golangci/golangci-lint golangci-lint run
```
Comment (rojas-diego) [inline Makefile:39]: You deleted the `.golangci.yml` file so maybe this rule doesn't work anymore?

File: Makefile [hunk 8]
```diff
@@ -1,34 +1,49 @@
...
+
+########### DataBase rules ##########
+
+run-db:
+	docker run --name accounts-mongo --detach --publish 27017:27017 mongo
+
+stop-db:
+	docker kill accounts-mongo
```
Comment (rojas-diego) [inline Makefile:47]: Your container should be named `notes-mongo`.

File: blocks.go [hunk 9]
```diff
@@ -2,191 +2,32 @@ package main
...
-	//objID, err := primitive.ObjectIDFromHex(in.Note_id)
-	var note_id = "62795c015f122130bb469849"
-
-	objID, err := primitive.ObjectIDFromHex(note_id)
-	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Error : during get note 1")
-	}
-
-	var note *notespb.Note
+type blocksService struct {
```
Comment (rojas-diego) [inline blocks.go:14]: Why did you call this `blocksService` if it's a `NotesAPI`?

**Review by rojas-diego — CHANGES_REQUESTED** [review]
Check Discord, I think it's better we take a different approach.

File: models/mongo/blocks.go (hunk 1 above)
Comment (rojas-diego) [reply models/mongo/blocks.go:20]: Same question here!

File: models/mongo/blocks.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type block struct {
+	ID      string `json:"id" bson:"_id,omitempty"`
+	NoteId  string `json:"noteId" bson:"noteId,omitempty"`
+	Type    uint32 `json:"type" bson:"type,omitempty"`
+	Content string `json:"content" bson:"content,omitempty"`
+}
```
Comment (rojas-diego) [inline models/mongo/blocks.go:18]: Why does this exist both in the models and the models.mongo package? Is there any reason for the duplication?

File: models/mongo/notes.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type note struct {
+	ID       string         `json:"id" bson:"_id,omitempty"`
+	AuthorId string         `json:"authorId" bson:"authorId,omitempty"`
+	Title    *string        `json:"title" bson:"title,omitempty"`
+	Blocks   []models.Block `json:"blocks" bson:"blocks,omitempty"`
+}
```
Comment (rojas-diego) [inline models/mongo/notes.go:18]: What's the reasoning behind having this `note` structure within the `mongo` package? Can't it be defined within the `models` package and then used by `mongo`?

File: models/note.go [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+	"github.com/google/uuid"
+)
+
+type NoteWithBlocks struct {
+	ID       uuid.UUID `json:"id" bson:"_id,omitempty"`
+	AuthorId string    `json:"authorId" bson:"authorId,omitempty"`
+	Title    string    `json:"title" bson:"title,omitempty"`
+	Blocks   []Block   `json:"blocks" bson:"blocks,omitempty"`
+}
```
Comment (rojas-diego) [inline models/note.go:16]: Is there such a thing as a note without blocks? What is the purpose of this type?

File: models/note.go [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+type NoteFilter struct {
+	ID       uuid.UUID `json:"id" bson:"_id,omitempty"`
+	AuthorId string    `json:"authorId" bson:"authorId,omitempty"`
+}
```
Comment (rojas-diego) [inline models/note.go:21]: Same thing as with the blocks. Using this filter type, you can either target one note by ID or a group of notes (given a user can be the author of multiple notes). I think it's better to have a `NoteFilter` and `ManyNotesFilter` and to differentiate `Delete`, `Get`, `Update` operations (which delete, get and update one single note) and `DeleteMany`, `List` and `UpdateMany` operations. I'm open to talk discuss over Discord if you want.

File: models/note.go [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+// NotesRepository is safe for use in multiple goroutines.
+type NotesRepository interface {
+	Create(ctx context.Context, noteRequest *NoteWithBlocks) (*NoteWithBlocks, error)
+
+	Get(ctx context.Context, filter *NoteFilter) (*NoteWithBlocks, error)
+
+	Delete(ctx context.Context, filter *NoteFilter) error
+
+	Update(ctx context.Context, filter *NoteFilter, noteRequest *NoteWithBlocks) error
```
Comment (rojas-diego) [inline models/note.go:31]: A call to `Update` should return the updated representation of the object according to REST principles.
```suggestion
	Update(ctx context.Context, filter *NoteFilter, note *NoteWithBlocks) (*NoteWithBlocks, error)
```

File: main.go [hunk 15]
```diff
@@ -1,72 +1,38 @@
...
-
-	if err != nil {
-		defer client.Disconnect(ctx)
-		log.Fatal(err)
-	}
-
-	NotesDatabase = client.Database("Notes-database")
-	NotesCollection = NotesDatabase.Collection("Notes-collection")
+	NotesDatabase   *mongo.Database   = nil
+	NotesCollection *mongo.Collection = nil
```
[the remark this answers is not in the document]
Comment (rojas-diego) [reply main.go:24]: Just realised you have imported some code from the accounts service for that! I guess this is dead code then?

File: server.go [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+		s.logger, err = zap.NewDevelopment(zap.AddStacktrace(zapcore.FatalLevel), zap.WithCaller(false))
+	}
+	must(err, "unable to instantiate zap.Logger")
+	s.slogger = s.logger.Sugar()
+}
+
+func (s *server) initAuthService() {
+	rawKey, err := base64.StdEncoding.DecodeString(*jwtPrivateKey)
+	must(err, "could not decode jwt private key")
+	s.authService = auth.NewService(rawKey)
```
Comment (rojas-diego) [inline server.go:109]: Looks like you're providing a private key to `auth.NewService` which expects a public key. Happy to discuss it through Discord if you want me to clarify.

**Review by rojas-diego — CHANGES_REQUESTED** [review]
Finished reviewing everything! Neat changes. I just have a few remarks on the implementation of the `models` package, I think it's easier if we hop on a call to discuss those! LMK :)

**Pull request merged**